### DISCUSSION TITLE
Mercure optimization: Suspend or share EventSource between tabs?

### DISCUSSION THREAD

**Discussion by stephanvierkant** [symfony/symfony#43603] [Q&A]
I've got a Symfony application with API Platform and Mercure. In my server logs I discovered a lot of connections to Mercure, even with very little users. Then I found out that when a user opens 5 tabs with the exact same page, there are 5 connections between that single user and our Mercure server. And some users had more than 10 tabs with the exact same page. Having this many connections is a waste of resources (for both my server and the end user).

I thought of two ideas:
1. Share a SSE connection between tabs. I found questions about sharing [them between tabs](https://stackoverflow.com/questions/32277583/is-there-a-way-to-have-one-server-sent-event-running-per-user-not-per-tab) or [using a ShareWorker](https://stackoverflow.com/questions/32321295/how-to-make-eventsource-available-inside-sharedworker-in-firefox) but they didn't give an (easy) answer. [One answer](https://stackoverflow.com/a/61083924/1405981) says it isn't possible, but I also found a proof of concept: [SSEGWSW](https://github.com/shoom3301/ssegwsw) ([blog](https://medium.com/its-tinkoff/ssegwsw-server-sent-events-gateway-by-service-workers-6212c1c55184)).

2. Pause/suspending connections in inactive tabs. I thought this would be very simple, but I can't find out how. Of course I can use `visibilitychange`, but that means that tab will miss messages and I'll to think of a way to fetch data when the tab becomes visible again.

I thought I wouldn't be the only one with this problem, but I can't find that many results. What are your thoughts on this? What's the best way to optimize this?

**Comment by norkunas** [comment]
Keep in mind that browsers limits how many SSE connections it can open from same source :)

Also for discussion - what is not working for you with a SharedWorker? I've tested it now and looks like it works, at least locally :)

**Reply by stephanvierkant** [reply]
I know there is a limit, so even more reason to share connections between tabs.

In the examples I've mentioned, the EventSource url is hardcoded in the worker. I'm looking for a solution that url can be specific (per user, but also multiple per page).

**Reply by norkunas** [reply]
I think you could post the message from context to worker with a URL you need to subscribe, and maintain a registry of hubs in your case if you'd send different URL's
